File: mojo/common/trace_provider_impl.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/common/trace_provider_impl.h"

#include <utility>

#include "base/callback.h"
#include "base/logging.h"
#include "base/memory/weak_ptr.h"
#include "base/time/time.h"
#include "base/trace_event/trace_config.h"
#include "base/trace_event/trace_event.h"
#include "mojo/public/cpp/application/application_impl.h"

namespace mojo {

TraceProviderImpl::TraceProviderImpl()
    : binding_(this), tracing_forced_(false), weak_factory_(this) {}

TraceProviderImpl::~TraceProviderImpl() {}

void TraceProviderImpl::Bind(InterfaceRequest<tracing::TraceProvider> request) {
  if (!binding_.is_bound()) {
    binding_.Bind(request.Pass());
  } else {
    LOG(ERROR) << "Cannot accept two connections to TraceProvider.";
  }
}

void TraceProviderImpl::StartTracing(
    const String& categories,
    mojo::InterfaceHandle<tracing::TraceRecorder> recorder) {
  DCHECK(!recorder_.get());
  recorder_ = tracing::TraceRecorderPtr::Create(std::move(recorder));
  tracing_forced_ = false;
  if (!base::trace_event::TraceLog::GetInstance()->IsEnabled()) {
    std::string categories_str = categories.To<std::string>();
    base::trace_event::TraceLog::GetInstance()->SetEnabled(
        base::trace_event::TraceConfig(categories_str,
                                       base::trace_event::RECORD_UNTIL_FULL),
        base::trace_event::TraceLog::RECORDING_MODE);
  }
}

void TraceProviderImpl::StopTracing() {
  DCHECK(recorder_);
  base::trace_event::TraceLog::GetInstance()->SetDisabled();

  base::trace_event::TraceLog::GetInstance()->Flush(
      base::Bind(&TraceProviderImpl::SendChunk, base::Unretained(this)));
}

void TraceProviderImpl::ForceEnableTracing() {
  base::trace_event::TraceLog::GetInstance()->SetEnabled(
      base::trace_event::TraceConfig("*", base::trace_event::RECORD_UNTIL_FULL),
      base::trace_event::TraceLog::RECORDING_MODE);
  tracing_forced_ = true;
  base::MessageLoop::current()->PostTask(
      FROM_HERE,
      base::Bind(&TraceProviderImpl::DelayedStop, weak_factory_.GetWeakPtr()));
}

void TraceProviderImpl::DelayedStop() {
  // We use this indirection to account for cases where the Initialize app
  // method (within which TraceProviderImpl is created) takes more than one
  // second to finish; thus we start the countdown only when the current thread
  // is unblocked.
  base::MessageLoop::current()->PostDelayedTask(
      FROM_HERE,
      base::Bind(&TraceProviderImpl::StopIfForced, weak_factory_.GetWeakPtr()),
      base::TimeDelta::FromSeconds(1));
}

void TraceProviderImpl::StopIfForced() {
  if (!tracing_forced_) {
    return;
  }
  base::trace_event::TraceLog::GetInstance()->SetDisabled();
  base::trace_event::TraceLog::GetInstance()->Flush(
      base::Callback<void(const scoped_refptr<base::RefCountedString>&,
                          bool)>());
}

void TraceProviderImpl::SendChunk(
    const scoped_refptr<base::RefCountedString>& events_str,
    bool has_more_events) {
  DCHECK(recorder_);
  // The string will be empty if an error eccured or there were no trace
  // events. Empty string is not a valid chunk to record so skip in this case.
  if (!events_str->data().empty()) {
    recorder_->Record(mojo::String(events_str->data()));
  }
  if (!has_more_events) {
    recorder_.reset();
  }
}

}  // namespace mojo

File: mojo/common/dart/lib/src/trace_provider_impl.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:async';

import 'package:mojo/core.dart';
import 'package:mojo_services/tracing/tracing.mojom.dart';

enum TraceSendTiming {
  IMMEDIATE,
  // TODO: Add BATCHED?
  AT_END,
}

class TraceProviderImpl implements TraceProvider {
  // Any messages sent before the tracing service connects to us will be
  // recorded and kept until one second after construction of the trace
  // provider. If the tracing service connects before that time, we will replay
  // the recorded trace events.
  //
  // This allows the client to record trace events early during initialization
  // of the app.
  List<String> _message_queue;
  bool _enqueuing;

  TraceProviderStub _stub;
  TraceRecorderProxy _recorder;
  // TODO(rudominer) We currently ignore _categories.
  String _categories;

  TraceSendTiming _timing;

  TraceProviderImpl([TraceSendTiming timing = TraceSendTiming.IMMEDIATE]) {
    _message_queue = [];
    _enqueuing = true;
    _timing = timing;
    new Future.delayed(const Duration(seconds: 1), () {
      if (_enqueuing) {
        _enqueuing = false;
        _message_queue.clear();
      }
    });
  }

  void connect(MojoMessagePipeEndpoint e) {
    _stub = TraceProviderStub.newFromEndpoint(e);
    _stub.impl = this;
  }

  @override
  void startTracing(String categories, TraceRecorderProxy recorder) {
    assert(_recorder == null);
    _recorder = recorder;
    _categories = categories;
    _enqueuing = false;
    if (_timing == TraceSendTiming.IMMEDIATE) {
      for (String message in _message_queue) {
        _recorder.ptr.record(message);
      }
      _message_queue.clear();
    }
  }

  @override
  void stopTracing() {
    assert(_recorder != null);
    if (_timing == TraceSendTiming.AT_END) {
      for (String message in _message_queue) {
        _recorder.ptr.record(message);
      }
      _message_queue.clear();
    }
    _recorder.close();
    _recorder = null;
  }

  bool isActive() {
    return _enqueuing || _recorder != null;
  }

  void sendTraceMessage(String message) {
    switch (_timing) {
      case TraceSendTiming.IMMEDIATE:
        if (_recorder != null) {
          _recorder.ptr.record(message);
        } else if (_enqueuing) {
          _message_queue.add(message);
        }
        break;
      case TraceSendTiming.AT_END:
        if (isActive()) {
          _message_queue.add(message);
        }
        break;
    }
  }
}
